### PR TITLE
feat: Refactored to a Plugin(model plugins) based Provider Architecture, and upgrade the Configure and Stream, session APIs

### PR DESCRIPTION
**Pull request by Nirupon123**
***BackGround***:
Previously, Susi Translator relied on a monolithic architecture where ML models and API keys were hardcoded globally. To achieve true multi-tenant isolation and open the door for community contributions, we needed a massive architectural overhaul. The primary goal of this PR is to make adding new AI models incredibly easy via a plugin system, while simultaneously providing the REST infrastructure our new frontend plugin needs to configure sessions and stream live captions.

- fixes: 
  #59  
- and completes,
  #57

- **Major Development**:
    - **Extensible Plugin Based Models**:
    This PR completely refactors how models are handled by introducing the ProviderRegistry and a pure plugin-based architecture.
         1. ***Developer Friendly***:   
             Anyone can now easily add a new model simply by writing an isolated Python file in the providers/plugins directory and registering it using __init___.py. 
         2. ***Mixed models***:
             Organizers can now dynamically pair entirely different plugins together on the fly.
         3. ***Lazy Loading & Isolation***:
             Heavy models are strictly isolated per tenant and only loaded into RAM when explicitly requested, preventing server-killing memory leaks.

- **Major Changes:**
   - Dynamic Configuration Endpoint(`/configure`):
      Upgraded the endpoint will act as the bridge between the frontend and the new plugin registry. It now accepts separate transcription and translation provider blocks. It dynamically provisions the exact plugins requested for that tenant and safely spins up the isolated audio grabber subprocesses.   
   - Real-time Server Sent Events Streaming (`/stream`):
     Added a highly efficient Server-Sent Events (SSE) endpoint. Instead of the frontend constantly polling the server, this endpoint pushes live, synced captions and translations directly to the browser as soon as they are generated by the plugins.
 
- **Other Important Changes:** 
   - ***Warmup Status*** (`/status`): Added a polling endpoint so the frontend loading screen knows exactly when heavy local models (like NLLB) have finished loading into RAM. This will be used for the loading screen between configuration and  streaming room.
   - ***Lifecycle Teardown***(`/stop_event`): Added a teardown endpoint to cleanly kill audio grabber subprocesses and completely wipe heavy ML plugins from memory when a session ends. 
   - ***Expanded Media Support***: Added twitch.tv and vimeo.com to the allowed host list in audio_sources.py.  
   - ***Test Suite Alignment***: Updated the pytest suite to cover the new split-pipeline payloads and gracefully handle the removed legacy code. 
 

## How To Test :
1. Start The Server :
```bash
uv run python transcribe_server.py
```
2. Create a New Session
```bash
   RES=$(curl -s -X POST http://localhost:5040/session \
     -H "Content-Type: application/json" \
     -d '{"source": "youtube"}')
      TENANT_ID=$(echo $RES | python3 -c "import sys, json; print(json.load(sys.stdin)['tenant_id'])")
      echo "Created Session with Tenant ID: $TENANT_ID"
 ```
 3. Configure Providers & Start the Audio Grabber
```bash
curl -s -X POST http://localhost:5040/api/v1/translate/configure \
     -H "Content-Type: application/json" \
     -d "{
           \"tenant_id\": \"$TENANT_ID\",
           \"transcription\": {
               \"provider_name\": \"whisper_local\",
               \"model_size\": \"small\"
           },
           \"translation\": {
               \"provider_name\": \"nllb_local\"
           },
           \"stream_url\": \"<Stream_URL>\"
         }"
```
**Note: Use any Vimeo, Twitch based Live Stream URL for now.**

4. Check Model Loading Status (Optional):
```bash
curl -s http://localhost:5040/api/v1/translate/status/$TENANT_ID
``` 

5.  Stream the Transcripts and Translations:
```bash
curl -N -s "http://localhost:5040/api/v1/translate/stream?tenant_id=$TENANT_ID&target_lang=<SET_TARGET_LANG>"
``` 

6. Tear Down the Session
```bash
curl -s -X POST http://localhost:5040/stop_event/$TENANT_ID
```

7. All Tests are Passing :
image

### CODE REVIEW

**Comment by Nirupon123** [comment]
@codex review

File: flask/transcribe_server.py [hunk 1]
```diff
@@ -261,25 +192,10 @@ def process_audio():
...
-            else:
-                # Local-model branch: torch was already imported at module
-                # load time when use_whisper_server=False, so this is cheap.
-                import torch 
-                model = model_fast if qsize > 20 else model_smart
-                audio_tensor = torch.from_numpy(audio_float32)
-                result = model.transcribe(audio_tensor, temperature=0)
-
-            transcript = (result.get('text') or '').strip()
+            transcript = registry.transcribe(tenant_id, audio_float32)
```
[the remark this answers is not in the document]
Comment (Nirupon123) [reply flask/transcribe_server.py:191]: not relevant right now , we can think about this in follow up PRs

File: flask/providers/registry.py [hunk 2]
```diff
@@ -8,93 +9,356 @@
...
+        #performing Atomic Writes to the registry with thread lock to avoid race conditions
+        with self._lock:
+            # always reset both slots on reconfigure to avoid stale provider bleedover
+            self._tenants[tenant_id] = {"transcription": None, "translation": None}
+
+            if transcription:
+                self._tenants[tenant_id]["transcription"] = {
+                    "provider_name": t_name,
+                    "config": transcription.get("config", {}),
+                    "instance": None,
```
[the remark this answers is not in the document]
Comment (Nirupon123) [reply flask/providers/registry.py:151]: resolved

**Review by Saksham-Sirohi — CHANGES_REQUESTED** [review]
Please fix the following before we move forward:
1. _xor_mask is duplicated and gives a false sense of security
2._resolve_instance has a thread-safety gap in the fallback path
3. Global SSL monkey-patch is dangerous
4._warmup calls private _load_model() by duck-typing
5. subprocess.Popen + os.setsid for audio_grabber needs cleanup on server shutdown
6. SSE /stream endpoint has no backpressure or client disconnect detection
7. pytest is a runtime dependency in pyproject.toml
8.is_available() in GroqWhisperProvider creates a client on every call; I don't think this is ideal

These are a few core items I feel require improvement here, ptal

**Comment by Nirupon123** [comment]
@Saksham-Sirohi addressed requested changes

**Review by weshaan — CHANGES_REQUESTED** [review]
Tested session creation, plugin configuration, SSE streaming, model lifecycle, and multi-tenant isolation. They work fine 👍 


----

#### 1. `stream_url` passed directly to subprocess without validation in `transcribe_server.py`, `configure_provider()`

```python
cmd = [sys.executable, "audio_grabber.py", "--tenant", tenant_id, "youtube", "--url", stream_url]
proc = subprocess.Popen(cmd, cwd=..., preexec_fn=os.setsid)
```

`stream_url` comes straight from the JSON body and is never validated before being passed as a subprocess argument. While `shell=False` prevents shell injection, the URL still reaches `yt-dlp` unvalidated,  `YouTubeSource._validate_url()` exists in `audio_sources.py` but is never called here. You can call it (or at minimum validate against `_ALLOWED_URL_SCHEMES` + `_ALLOWED_HOSTS`) before building the command.

---

#### 2. `tenant_id` used in `stop_event` path without any validation

```python
@app.route('/stop_event/<tenant_id>', methods=['POST'])
def stop_event(tenant_id):
    ...
    os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
```

Any caller who knows (or guesses) a `tenant_id` can kill another tenant's subprocess. There is zero authentication. This is by design for a prototype, but it's worth noting `tenant_id` is not a secret (it's returned in the session response), so this endpoint is effectively unauthenticated process termination.

**Comment by Nirupon123** [comment]
@weshaan authentication is handled in pr #84 , auth is not a scope of this pr.
You can check out, if you remove an allowed host from the list, it will not work on that scenario, that's why I added twitch and , vimeo in allowed hosts.

**Review by Saksham-Sirohi — COMMENTED** [review]
1. stream_url not validated at the API boundary

configure_provider() passes stream_url straight into subprocess.Popen → audio_grabber.py → YouTubeSource._validate_url(). Validation only happens in the child process.

Fail fast in the parent with YouTubeSource._validate_url(stream_url) (or equivalent host/scheme checks).
Avoids spawning processes for bad input and closes the gap [@weshaan flagged](https://github.com/fossasia/susi_translator/pull/83/).
shell=False avoids shell injection; the remaining risk is yt-dlp fetching attacker-chosen URLs (allowed-host bypass via redirects is a known class of issue — worth a quick check in audio_sources.py).

2. Global SSL monkey-patch in whisper_local.py

During model download, the code replaces urllib.request.urlopen process-wide and enables OP_LEGACY_SERVER_CONNECT. That weakens TLS for the entire Flask process, not just the download. Scope this to the download call (custom opener/context) and restore afterward.

3. Resource exhaustion (DoS)

No limits on:

concurrent /configure calls → subprocess + model warmup threads
concurrent SSE /stream clients → infinite while True loops doing translation work
auto-fallback to whisper_local / nllb_local when a tenant has no config → surprise RAM use
For anything beyond localhost, you need caps (max tenants, max grabbers, max SSE clients) or auth-gated quotas from #84.

4. Reconfigure leaks grabber processes

Calling /configure again with stream_url overwrites grabber_processes[tenant_id] without killing the old process → orphaned yt-dlp/ffmpeg children. Kill the existing grabber before spawning a new one.



you may have to implemement these in fllow up PRs

**Comment by Nirupon123** [comment]
@Saksham-Sirohi added an patch for the requested changes, and for the Rate limiting i didnt enforce here it will be done in a follow up PR after the auth pr

**Review by Saksham-Sirohi — APPROVED** [review]
ok, this now makes ample sense; further can be done in the auth pr

**Review by weshaan — APPROVED** [review]
SSL has Two problems: (a) it's a permanent process-wide replacement, original_urlopen is never restored, so every urllib.request.urlopen call for the life of the Flask process gets the weakened context; (b) OP_LEGACY_SERVER_CONNECT re-enables unsafe legacy renegotiation. This downgrades TLS for the whole app, not just the model download. Fix: build a custom ssl.SSLContext + opener, use it only for the download call, and never touch the global. Better still, hand the context to whisper's downloader directly rather than patching urlopen at all.

No authentication + client-controlled tenant_id = full IDOR (high risk). Every endpoint trusts a client-supplied tenant_id. POST /stop_event/<tenant_id> will kill any other tenant's grabber and delete their transcripts; /configure, /stream, and the /transcripts* family let any caller read or overwrite any tenant's data by guessing/enumerating IDs. Session IDs are uuid4().hex (good), but nothing binds a caller to the session. 

to be addressed in future PRs

  python -m pytest -q   →   127 passed in 3.14s
  
  python -m pytest -v -k "session or configure or stream or status or stop or isolation or warmup"
  → 40 passed
  
  Each test name maps to a guarantee: TestSession::*, TestConfigure::test_per_tenant_isolation_in_registry, TestStream::*, TestStatus::*, TestStop::*.

  
  SESSION            → unique tenant_id
 CONFIGURE       → status:success, whisper_local registered
  STATUS             → warming_up
  EDGE1 unknown provider→ 400 
  EDGE2 missing tenant  → 400
  EDGE3 stream no config→ HTTP 200
STOP               → success
  
  
image

**Comment by Nirupon123** [comment]
@weshaan patched the SSL issue in last push and again auth is not a scope of this pr if you are concerned about rate limiting then it will be enforced in follow up PRs
